src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -0,0 +1,566 @@
+        }
+
+        private static long ReadScatterAtOffset(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers, long fileOffset)
+        {
+            long total = 0;
+
+            // ReadFileScatter does not support sync handles, so we just call ReadFile in a loop
+            for (int i = 0; i < buffers.Count; i++)
+            {
+                Span<byte> span = buffers[i].Span;
+                int read = ReadAtOffset(handle, span, fileOffset + total);
+                total += read;
+
+                // We stop on the first incomplete read.
+                // Most probably there is no more data available and the next read is going to return 0 (EOF).
+                if (read != span.Length)
+                {
+                    break;
+                }
+            }
+
+            return total;
+        }
+
+        private static long WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<ReadOnlyMemory<byte>> buffers, long fileOffset)
+        {
+            long total = 0;
+
+            // WriteFileGather does not support sync handles, so we just call WriteFile in a loop
+            for (int i = 0; i < buffers.Count; i++)
+            {
+                ReadOnlySpan<byte> span = buffers[i].Span;
+                int written = WriteAtOffset(handle, span, fileOffset + total);
+                total += written;
+
+                // We stop on the first incomplete write.
+                // Most probably the disk became full and the next write is going to throw.
+                if (written != span.Length)
+                {
+                    break;
+                }
+            }
+
+            return total;
+        }
+
+        private static ValueTask<long> ReadScatterAtOffsetAsync(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers,
+            long fileOffset, CancellationToken cancellationToken)
+        {
+            if (CanUseScatterGatherWindowsAPIs(handle))
+            {
+                long totalBytes = 0;
+                for (int i = 0; i < buffers.Count; i++)
+                {
+                    totalBytes += buffers[i].Length;
+                }
+
+                if (totalBytes < int.MaxValue) // the ReadFileScatter API uses int, not long
+                {
+                    return ReadScatterAtOffsetSingleSyscallAsync(handle, buffers, fileOffset, cancellationToken);
+                }
+            }
+
+            return ReadScatterAtOffsetMultipleSyscallsAsync(handle, buffers, fileOffset, cancellationToken);
+        }
+
+        // From https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-readfilescatter:
+        // "The file handle must be created with the GENERIC_READ right, and the FILE_FLAG_OVERLAPPED and FILE_FLAG_NO_BUFFERING flags."
+        // "This function is not supported for 32-bit applications by WOW64 on Itanium-based systems."
+        private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
```

Comment:
Where is this checked?
"Each buffer must be at least the size of a system memory page and must be aligned on a system memory page size boundary."

Reply:
both `ReadFile` and `ReadFileScatter` require it for handles opened for `FILE_FLAG_OVERLAPPED` and `FILE_FLAG_NO_BUFFERING` and fail with the same error if it's not provided. So we don't have to check that.

Reply:
But isn't the intent of the API that it only uses the scatter/gather native functions if possible, and otherwise falls back to the implementation that always works?

Reply:
> otherwise falls back to the implementation that always works

the other implementation is also not going to work. Both `ReadFile` and `ReadFileScatter` have the same alignment requirements for `FILE_FLAG_NO_BUFFERING` handles.

Reply:
Ok.  As long as everything that can work does work regardless of implementation detail, I'm happy.